<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Comparing microbial communities with OptiFit: a fast method for fitting sequences to existing OTUs</title>
    <meta charset="utf-8" />
    <meta name="author" content="Kelly Sovacool" />
    <meta name="date" content="2021-03-11" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <script src="libs/clipboard/clipboard.min.js"></script>
    <link href="libs/shareon/shareon.min.css" rel="stylesheet" />
    <script src="libs/shareon/shareon.min.js"></script>
    <link href="libs/xaringanExtra-shareagain/shareagain.css" rel="stylesheet" />
    <script src="libs/xaringanExtra-shareagain/shareagain.js"></script>
    <link rel="stylesheet" href="xaringan-themer.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Comparing microbial communities with OptiFit: a fast method for fitting sequences to existing OTUs
### Kelly Sovacool
### 2021-03-11

---












background-image: url("../../figures/SchlossLab.png") 
background-size: contain
background-color: "#000000"
class: inverse

---

## Role of the microbiome in human health

.center[
&lt;img src="https://raw.githubusercontent.com/kelly-sovacool/bioinf603-fall2019/master/docs/figures/human_microbiome.gif" height="500"/&gt;
]

---

## The gut microbiome changes in colorectal cancer

.pull-left[
- Colorectal cancer patients have an altered gut microbiome compared to healthy individuals.

- Fecal matter transplants from cancerous mice increase tumor formation in germ-free mice.

- Machine learning models trained on microbiome data to predict CRC have modest performance (mean AUROC ~0.7) 
]

.pull-right[
![](../../figures/zackular-3a.jpg)
]

.footnote[
[Zackular et al. (2013) _mBio_](doi.org/10.1128/mBio.00692-13)
]

---

## Characterizing microbial communities

.pull-left[
&lt;img src="../../figures/omics-cascade.jpg" height=500/&gt;
]

--

.pull-right[

- Genomics: Which microbes are there &amp; what are they capable of doing

- Transcriptomics: What are they trying to do

- Proteomics: What are they actually doing

- Metabolomics: How are they affecting the host &amp; each other

.footnote[
[Haukaas et al. (2017) _Metabolites_](doi.org/10.3390/metabo7020018)
]
]

---

## Sequencing techniques

|   | Metagenomics | Amplicon Sequencing |
|------|--------------|----------|
| **Technique** | Whole-metagenome shotgun sequencing: every gene of every microbe | Sequence just 250bp of the 16S rRNA gene |
| **Cost** | ðŸ’²ðŸ’²ðŸ’² | ðŸ’² |
| **Taxonomic resolution** | Species or strain | Genus |
| **Organisms** | Bacteria, archaea, viruses, fungi | Bacteria &amp; archaea; or fungi separately | 
| **Purpose** | Identify which microbial genes are present; assemble microbial genomes | Identify which microbes are present and their abundances |

???
metagenomics is prohibitively expensive for very large datasets

---

## Operational Taxonomic Units (OTUs)

- Cluster similar sequences together into OTUs as a proxy for genus/species.
- OTUs clustered at 3% distance (97% sequence similarity) generally can be labelled at the genus level.

.center[
&lt;img src="https://raw.githubusercontent.com/kelly-sovacool/bioinf603-fall2019/master/docs/figures/OTU_clustering.png" height="400" /&gt;
]
.pull-right[Image credit: M. Brodie Mumphrey]

---

## Why cluster sequences into OTUs?

- Bacterial taxonomy is imperfect. Some species are overly broad, others are too narrow.
  - e.g. some strains of _E. coli_ share only 20% of their genomes.
  
- Many bacteria have multiple copies of the rRNA gene.
  - Pick a sequence distance threshold to minimize splitting sequences from the same genome into different OTUs.
  
- Bacteria evolve rapidly. Don't want to exclude novel microbes from an analysis.

???

Bacterial taxonomy is hard

Goal of classification is to reflect evolutionary history.

Challenges:
- Asexual reproduction
- Fast evolutionary change
- Horizontal gene transfer

Current classifications are far from perfect.

Courtney's resolution figure?

---

## Evaluating OTU Quality

Pairs of sequences that are under the distance threshold should be assigned to the same OTU; sequences above the threshold should be assigned to different OTUs.

Confusion Matrix:

|   | Actual Positive | Actual Negative |
|---|---|---|
| Predicted Positive | TP | FP |
| Predicted Negative | FN | TP |

--

Matthews Correlation Coefficient:

`$$MCC = \frac{TP \times TN - FP \times FN}{\sqrt{(TP + FP)(TP + FN)(TN + FP)(TN + FN)}}$$`

Range of MCC:

- 1 - perfect prediction
- 0 - random
- -1 - completely wrong

---

## OTU Clustering methods

- _De novo_ 
  - Cluster sequences based on their similarity to each other.
  - Limitation: OTUs aren't stable; you'll get different OTU assignments when new sequences are added.
- Reference-based 
  - Cluster sequences based on their similarity to a reference database.
  - Limitation: novel sequences won't be represented. OTUs are limited by the diversity represented in the database.

.center[
&lt;img src="../../figures/schloss-2016-msystems_all-method-comparison.png"  height="350"/&gt;
]

???

_De novo_ methods out-perform reference-based methods.

.footnote[

  [Westcott &amp; Schloss (2015) _PeerJ_ ](https://doi.org/10.7717%2Fpeerj.1487);
  [Schloss (2016) _mSystems_ ](https://doi.org/10.1128%2FmSystems.00027-16);
  [Westcott &amp; Schloss (2017) _mSphere_](10.1128/mSphereDirect.00073-17)
]

---

## OptiClust: _de novo_ OTUs that optimize the MCC

1. Calculate all pairwise sequence distance scores.

1. Begin with each sequence assigned to its own OTU.

1. For each sequence, consider whether to move it to a different OTU or not. Do whichever option results in the best MCC score. Repeat until the MCC doesn't change.

.center[
&lt;img src="../../figures/opticlust-supp-ex.png" width="700"/&gt;
]
.footnote[
&lt;img src="https://raw.githubusercontent.com/mothur/logo/master/mothur_RGB.png" width="200"/&gt;
]
???
if equally good, pick one at random

---

## OptiFit: assign new sequences to existing OTUs

1. Start with OTUs that were previously clustered _de novo_ to use as a reference.

1. Calculate pairwise sequence distance scores between all reference and query sequences.

1. Begin with each query sequence randomly assigned to a reference OTU.

1. For each query sequence, consider whether to move it to a different OTU or not. Do whichever option results in the best MCC score. Repeat until the MCC doesn't change.

&amp;nbsp;  
&amp;nbsp;
&amp;nbsp;
&amp;nbsp;

.pull-left[
&lt;img src="https://raw.githubusercontent.com/mothur/logo/master/mothur_RGB.png" width="200"/&gt;
]

---

## Reference-based strategies

- Modes
  - Open - sequences that can't be fit to the reference are then clustered _de novo_ to create additional OTUs.
  - Closed - sequences that can't be fit to the reference are thrown out.
  
- Reference choice
  - External, independent database (Silva, greengenes, RDP).
  - Split the dataset into a reference &amp; query set.

---

## Reference clustering against a database

Questions

- Which database should we use?
  - SILVA, Greengenes, Ribosomal database project
  
- How does the performance of fitting datasets to independent databases compare to _de novo_ clustering?

- Are the results consistent across datasets from a variety of sources? 
  - Human gut, mouse gut, marine, soil
  
- How does mothur compare to vsearch?

---

## Reference clustering against a database

.pull-left[

&lt;img src='../../exploratory/2021/figures/mcc_fit-db-1.png' /&gt;
]

--

.pull-right[

&lt;img src='../../exploratory/2021/figures/fraction-mapped_fit-db-1.png' /&gt;
]

???
- _De novo_ and open-reference perform equally well.
- Closed reference performs slightly worse.
- RDP doesn't perform well.
- Prior to switching to silva v132 &amp; updating mothur, closed-reference artificially outperformed open-reference.

---

## Reference clustering against a database

.pull-left[

&lt;img src='../../exploratory/2021/figures/runtime_fit-db-1.png' /&gt;
]

--

.pull-right[

&lt;img src='../../exploratory/2021/figures/memory_fit-db-1.png' /&gt;
]

---


## mothur vs. vsearch


&lt;img src='../../exploratory/2021/figures/metrics_all-human-1.png' /&gt;

???
- Just showing the human dataset here because results patterns were basically the same across all datasets.
- Only used greengenes for reference because that's what
QIIME does.

---

## Splitting datasets into reference &amp; query sets

Questions

- What fraction of sequences should be in the reference?

- How to select which sequences to be in the reference?
  - Simple random sample
  - Weight by sequence abundance
  - Weight by connectivity (number of pairs within the distance threshold)
  
- How does the performance of fitting datasets to reference fractions compare to de novo clustering? 

- Are the results consistent across datasets from a variety of sources?
  - Human gut, mouse gut, marine, soil

---

## Splitting datasets into reference &amp; query sets

&lt;img src='../../exploratory/2020-05_May-Oct/figures/mcc_complement-4.png' /&gt;


---

## Splitting datasets into reference &amp; query sets

&lt;img src='../../exploratory/2021/figures/fraction-mapped_fit-split-1.png' /&gt;

---

## Splitting datasets into reference &amp; query sets

&lt;img src='../../exploratory/2021/figures/runtime_fit-split-1.png' /&gt;

---

## Summarizing mothur results

&lt;img src='../../exploratory/2021/figures/mcc_all-opti-1.png' /&gt;

???

Whenever there's virtually no variation in the metric across random seeds.

---
## Conclusions

- If you need to assign new sequences to existing OTUs, OptiFit results in OTUs of similar quality as OptiClust.

- _De novo_ clustering still usually outperforms reference clustering against an external database.

- VSEARCH runs faster but produces lower quality OTUs than OptiFit or OptiClust.

- OptiFit runs slower than OptiClust in some cases; continue using OptiClust if stable OTUs aren't required.

--

## What's next

- Demonstrate that OptiFit works well for machine learning.

- Integrate OTUs with other microbial omics data.

---

## Managing Bioinformatics Pipelines

.center[
&lt;img src="https://snakemake2019.readthedocs.io/en/latest/_images/workflow_management.png" width="750"/&gt;
]

.footnote[
[snakemake.readthedocs.io](snakemake.readthedocs.io)
]

???
I ran about 20,000 jobs on the GreatLakes HPC

---

## Pipelines with &lt;img src="https://raw.githubusercontent.com/snakemake/snakemake/f55801bfabfb916f77fa4ec8f9eb0667e5c8a92e/images/biglogo.png" height="35"/&gt; 

.center[
&lt;img src="https://raw.githubusercontent.com/snakemake/snakemake/f55801bfabfb916f77fa4ec8f9eb0667e5c8a92e/docs/project_info/img/idea.png" width="900" /&gt;
]

.footnote[
[snakemake.readthedocs.io](snakemake.readthedocs.io)
]

---

## Tools for Reproducible Bioinformatics 

| package mgmt | scripts | reports | workflow | version control |
|---|---|---|---|---|
|   &lt;img src="https://docs.conda.io/en/latest/_images/conda_logo.svg" width="175"/&gt; | &lt;img src="https://longhowlam.files.wordpress.com/2017/04/rp.png" height="175"/&gt; | &lt;img src="https://ulyngs.github.io/rmarkdown-workshop-2019/slides/figures/rmarkdown.png" height="165"/&gt; | &lt;img src="https://avatars.githubusercontent.com/u/33450111?s=400&amp;v=4" height="175"&gt; |&lt;img src="https://miro.medium.com/max/2625/1*8fPMdk2Cd5iJQ7dI7jXCbA.jpeg" height="175"/&gt; |

--

.center[
### ...and headache prevention!
]

---

## Sequence similarity / distance

1. Calculate all pairwise sequence identity scores to create a distance matrix.
2. Apply cutoff of 97% sequence identity to create an adjacency matrix.
.center[
```
seqA AGGGTACG
seqB ACG-TACG
```
$$ \%\ sequence\ identity =  100 * \frac{number\ of\ matches}{sequence\  length} $$
]
.pull-left[
1) Distance matrix:

| | Seq1 | Seq2 | Seq3 |
|-|------|------|------|
| Seq1 | - | 98 | 87 |
| Seq2 | 98 | - | 76 |
| Seq3 | 87 | 76 | - |
]
.pull-right[
2) Adjacency matrix:

| | Seq1 | Seq2 | Seq3 |
|-|------|------|------|
| Seq1 | - | 1 | 0 |
| Seq2 | 1 | - | 0 |
| Seq3 | 0 | 0 | - |
]
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="libs/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
