---
title: "**OptiFit: a fast method for fitting amplicon sequences to existing OTUs**"
date: '`r Sys.Date()`'
authors:
  - name: "Kelly L. Sovacool"
    url: https://github.com/kelly-sovacool
    affiliation: Department of Computational Medicine and Bioinformatics, University of Michigan, Ann Arbor, MI
    orcid_id: 0000-0003-3283-829X
  - name: "Sarah L. Westcott"
    url: https://github.com/mothur-westcott
    affiliation: Department of Microbiology and Immunology, University of Michigan, Ann Arbor, MI
  - name: "M. Brodie Mumphrey"
    url: https://github.com/MBMumphrey
    affiliation: Department of Computational Medicine and Bioinformatics, University of Michigan, Ann Arbor, MI
  - name: "Gabrielle A. Dotson"
    url: https://github.com/dotsonga
    affiliation: Department of Computational Medicine and Bioinformatics, University of Michigan, Ann Arbor, MI
  - name: "Patrick D. Schloss"
    url: https://github.com/pschloss
    affiliation: Department of Microbiology and Immunology, University of Michigan, Ann Arbor, MI
bibliography: references.bib
output:
  distill::distill_article:
    toc: true
    toc_float: true
    toc_depth: 2
  pdf_document:
    keep_tex: false
    includes:
      in_header: preamble.tex
      before_body: head.tex
      after_body: tail.tex
knit: (function(inputFile, encoding) {
  rmarkdown::render(inputFile, encoding = encoding, output_dir = "docs", output_format = "all") })
csl: msystems.csl
fontsize: 12pt
geometry: margin=1.0in
repository_url: https://github.com/SchlossLab/OptiFitAnalysis
creative_commons: CC BY-SA
twitter:
  creator: "@kelly_sovacool"
  site: "@PatSchloss"
---

```{r knitr_settings, eval=TRUE, echo=FALSE, cache=FALSE}
mothuR::set_knitr_opts()
```

## Abstract

### Importance

\newpage

## Introduction

## Results

- 16S rRNA gene sequence datasets from human gut, mouse gut, marine, and soil environments were downloaded, processed with mothur, and clustered _de novo_ with OptiClust.
- Independent reference databases (greengenes, silva, and RDP) were clustered _de novo_, then query datasets were fit to the _de novo_ OTUs using OptiFit.
  - In open-reference mode, OTU quality as measured by the MCC score was similar between fitting the datasets to reference OTUs with OptiFit and clustering the datasets _de novo_ with OptiClust. 
  - However, in closed-reference mode, OTU quality was slightly worse when fitting to greengenes and silva, and much worse when fitting to RDP as compared to OptiClust. OptiFit was able to map more query sequences to reference OTUs created with the greengenes and silva databases than with RDP.
  - These results held true for all four datasets and all three reference databases.
- Datasets were randomly split into a reference fraction and a query fraction.
  - Reference sequences were clustered _de novo_ with OptiClust.
  - Query sequences were then fit to the _de novo_ OTUs with OptiFit.
- vsearch with greengenes

## Discussion

## Materials and Methods

## Acknowledgements {.appendix}

KLS received support from the NIH Training Program in Bioinformatics (T32 GM070449).

The funders had no role in study design, data collection and interpretation, 
or the decision to submit the work for publication.

## Author Contributions {.appendix}
